libcudacxx/test/libcudacxx/std/language.support/support.srcloc/general.pass.cpp

```diff
@@ -0,0 +1,173 @@
+static_assert(empty.file_name()[0] == '\0', "");
+static_assert(empty.function_name()[0] == '\0', "");
+
+ASSERT_NOEXCEPT(empty.line());
+ASSERT_NOEXCEPT(empty.column());
+ASSERT_NOEXCEPT(empty.file_name());
+ASSERT_NOEXCEPT(empty.function_name());
+static_assert(cuda::std::is_same<unsigned, decltype(empty.line())>::value, "");
+static_assert(cuda::std::is_same<unsigned, decltype(empty.column())>::value, "");
+static_assert(cuda::std::is_same<const char*, decltype(empty.file_name())>::value, "");
+static_assert(cuda::std::is_same<const char*, decltype(empty.function_name())>::value, "");
+
+__device__ _CCCL_CONSTEXPR_GLOBAL cuda::std::source_location device_empty{};
+#if !defined(_CCCL_CUDACC_BELOW_11_3)
+static_assert(device_empty.line() == 0, "");
+static_assert(device_empty.column() == 0, "");
+static_assert(device_empty.file_name()[0] == '\0', "");
+static_assert(device_empty.function_name()[0] == '\0', "");
+#endif // _CCCL_CUDACC_BELOW_11_3
+
+ASSERT_NOEXCEPT(device_empty.line());
+ASSERT_NOEXCEPT(device_empty.column());
+ASSERT_NOEXCEPT(device_empty.file_name());
+ASSERT_NOEXCEPT(device_empty.function_name());
+
+// A simple use of current() outside a function.
+#line 1000 "ss"
+constexpr cuda::std::source_location cur = cuda::std::source_location::current();
+static_assert(cur.line() == 1000, "");
+
+#if _CCCL_HAS_BUILTIN(__builtin_COLUMN) || defined(TEST_COMPILER_MSVC) && _CCCL_MSVC_VERSION >= 1927
+static_assert(cur.column() > 0, "");
+#else // ^^^ _CCCL_BULTIN_COLUMN ^^^ / vvv !_CCCL_BULTIN_COLUMN vvv
+static_assert(cur.column() == 0, "");
+#endif // !_CCCL_BULTIN_COLUMN
+static_assert(cur.file_name()[0] == 's' && cur.file_name()[1] == 's' && cur.file_name()[2] == '\0', "");
+
+// MSVC below 19.27 is broken with function name
+#if !defined(_CCCL_COMPILER_MSVC) || _CCCL_MSVC_VERSION >= 1927
+static_assert(cur.function_name()[0] == '\0', "");
+#else // ^^^ __builtin_FUNCTION ^^^ / vvv !__builtin_FUNCTION vvv
+static_assert(compare_strings(cur.function_name(), "__builtin_FUNCTION is unsupported!"));
+#endif // !__builtin_FUNCTION
+
+__host__ __device__ bool compare_strings(const char* lhs, const char* rhs) noexcept
+{
+  for (size_t index = 0;; ++index)
+  {
+    if (lhs[index] != rhs[index])
+    {
+      return false;
+    }
+
+    if (lhs[index] == '\0')
+    {
+      return true;
+    }
+  }
+}
+
+__host__ __device__ bool find_substring(const char* source, const char* target) noexcept
+{
+  if (target[0] == '\0')
+  {
+    return true;
+  }
+
+  for (size_t index = 0;; ++index)
+  {
+    if (source[index] == target[index])
+    {
+      for (size_t sub_index = 0;; ++sub_index)
+      {
+        if (source[sub_index] != target[sub_index])
+        {
+          break;
+        }
+
+        if (target[sub_index] == '\0')
+        {
+          return true;
+        }
+      }
+    }
+
+    if (target[index] == '\0')
+    {
+      return false;
+    }
+  }
+}
+
+__host__ __device__ void test()
+{
+#line 2000
```

Comment:
This will also ruin line numbers below. Suggestion: forward declare a function such as:

```cpp
cuda::std::source_location fake_source_location();
```

Then call that function inside the unttest. At the VERY END of this file add this:

```cpp
#line 1000 "sss"
cuda::std::source_location fake_source_location() { return cuda::std::source_location::current(); }
// End of file
```

Then you have deterministic values for line and file.

Reply:
Ditto, it is testing intended functionality and if somethign breaks the test is short enough to work with it